mcs/class/System/System.Net.Sockets/SocketAsyncEventArgs.cs

```diff
@@ -131,13 +176,10 @@ void Dispose (bool disposing)
 			disposed = true;
 
 			if (disposing) {
-				if (disposed || Interlocked.CompareExchange (ref in_progress, 0, 0) != 0)
+				if (Interlocked.CompareExchange (ref in_progress, 0, 0) != 0)
```

Comment:
This is one weird way of doing a volatile read. Just mask in_progress as volatile.